### ISSUE TITLE
`CommandInteractionOptionResolver.getSubCommand()` should return `undefined` when a sub command isn't used

### ISSUE THREAD

**Issue by APixelVisuals** [discordjs/discord.js#6142]
**Is your feature request related to a problem? Please describe.**
[`CommandInteractionOptionResolver.getSubCommand()`](https://discord.js.org/#/docs/main/master/class/CommandInteractionOptionResolver?scrollTo=getSubCommand) currently throws an error if a sub command isn't used. I think it would be easier if it returned `undefined` and had a `required` parameter similar to [`CommandInteractionOptionResolver.get()`](https://discord.js.org/#/docs/main/master/class/CommandInteractionOptionResolver?scrollTo=get). The same applies for [`CommandInteractionOptionResolver.getSubCommandGroup()`](https://discord.js.org/#/docs/main/master/class/CommandInteractionOptionResolver?scrollTo=getSubCommandGroup).

**Describe the ideal solution**
Return `undefined` if `required` is `false`

**Describe alternatives you've considered**
Create a utility module that catches the error and returns `undefined`

**Comment by monbrey** [comment]
If this change is to be made it should return `null`, not `undefined`.

**Comment by shinotheshino** [comment]
According to [the docs](https://discord.com/developers/docs/interactions/slash-commands#nested-subcommands-and-groups), you can't have a sub-command/group defined and also use the top-level command. What is the use case for this? One of the benefits of `CommandInteractionOptionResolver` is validating options at runtime against the command data when defined and I don't see how this would really play into that.

**Comment by monbrey** [comment]
> According to [the docs](https://discord.com/developers/docs/interactions/slash-commands#nested-subcommands-and-groups), you can't have a sub-command/group defined and also use the top-level command.

You could have a global and a guild command with the same name, one with subcommands and one without.

I think that's an absolutely terrible design but it is theoretically possible.

**Comment by shinotheshino** [comment]
That could still be handled safely by `if (interaction.command.guildId)`, though, couldn't it?

**Comment by Jiralite** [comment]
All `<CommandInteractionOptionResolver>.getSubCommand()` does is retrieve the private property `._subCommand` and fails with a thrown error if does not exist. The same also applies to sub-command groups.

Why not just make both those private properties public? You could then remove the methods and instead make those properties nullable (which they already are). I don't see a need for these two methods to exist when they exist fine as properties.

I also ask this because I utilise a command logger where Slash Commands are logged in the raw format sent, and I really do not like having to `try {...} catch {...}` just to retrieve a variable dynamically. I also see people on the discord.js Discord server in disapproval of this behaviour!

If we were to change this, I would move that we remove the`<CommandInteractionOptionResolver>.getSubCommand()` and `<CommandInteractionOptionResolver>.getSubCommandGroup()` methods, rename `._group` to `._subcommandGroup` to keep it consistently named with the other property and finally make the two private properties public.

**Comment by shinotheshino** [comment]
One of the purposes of the resolver is to abstract away as many possibilities for bugs as possible for the user. With this intent in mind, throwing an error when attempting to access the sub-command of an interaction without one makes sense. For your logging use-case, a `try/catch` expression, checking the error code, is perfectly reasonable *and* explicit.

IMO, it makes no sense to add less strict code to the library just to accommodate such a case when there's a perfectly reasonable way to get it via `try/catch`.

**Comment by Jiralite** [comment]
> One of the purposes of the resolver is to abstract away as many possibilities for bugs as possible for the user. With this intent in mind, throwing an error when attempting to access the sub-command of an interaction without one makes sense.

Then, I don't think this should be the purpose of the resolver. Ideally, the resolver would have looked at the data received and parsed them into a more accessible format for the user rather than chaining through the `Collection` as before. Throwing exceptions on this seems a bit of a stretch when you can just return `null` and let users handle that behaviour as they please. It's also kind of inconsistent with the other methods due to being unable to suppress the error with an argument.

Though as I say this, if this issue isn't addressed, at least we have #6156 so we can ignore all the methods and choose to go into the options ourselves to handle what we want _how_ we want.

> IMO, it makes no sense to add less strict code to the library just to accommodate such a case when there's a perfectly reasonable way to get the data via try/catch.

This isn't just one case! But this sentence is only really valid if the strict code was a good idea in the first place!

**Comment by monbrey** [comment]
> Though as I say this, if this issue isn't addressed, at least we have #6156 so we can ignore all the methods and choose to go into the options ourselves to handle what we want _how_ we want.

This one will be merged to provide raw access, which sounds like it meets the use case you described of logging the raw format. We acknowledged and agreed that there were reasons people want the actual data rather than resolving things, and will be providing both.

What I'm still not seeing here is a use case where people _would_ want to be using the resolver, and trying to call `getSubCommand` when a sub-command doesn't exist. If a command has sub-commands, then one of the must have been used.

I think if you're making potentially incorrect calls like this in your code, getting a sub-command where one doesn't exist, then it's appropriate to be catching errors.

**Comment by Jiralite** [comment]
> This one will be merged to provide raw access

It doesn't seem raw to me - it looks to be processed! Though I'm fine with that.

> What I'm still not seeing here is a use case where people _would_ want to be using the resolver, and trying to call `getSubCommand` when a sub-command doesn't exist. If a command has sub-commands, then one of the must have been used.

One other thing is that I have my commands set up something like this as a `Collection` for my Slash Command handler:
```
Collection(3) [Map] {
  'command1' => Collection(3) [Map] {
    'sub_command1' => [sub_command1],
    'sub_command2' => [sub_command2],
    'sub_command3' => [sub_command3]
  },
  'command2' => Collection(3) [Map] {
    'sub_command1' => [sub_command1],
    'sub_command2' => [sub_command2],
    'sub_command_group1' => Collection(2) [Map] {
      'sub_command1' => [sub_command1],
      'sub_command2' => [sub_command2]
    }
  },
  'command3' => command3 { name: 'command3' }
}
```
In the `Collection`, the command names are what we first see. If a sub-command is used, those belong inside the respective commands. If a sub-command-group is present, it goes even deeper!

I structure my commands like this. When I'm dynamically running my Slash Commands, I have to check appropriately if what is being sent belongs to a sub-command group or if it's just a sub-command alone or just a plain ol' Slash Command. Upon receiving one, I traverse through the respective `Collection`s until I'm at the right place to execute my method. This is why I would always check for a sub-command group and a sub-command.
